clients/pkg/promtail/targets/docker/target.go

```diff
@@ -198,6 +198,15 @@ func (t *Target) process(r io.Reader, logStream string) {
 
 }
 
+// startIfNotRunning starts processing container logs. The operation is idempotent , i.e. the processing cannot be started twice.
+func (t *Target) startIfNotRunning() {
+	if t.running.CAS(false, true) {
+		ctx, cancel := context.WithCancel(context.Background())
+		t.cancel = cancel
```

Comment:
This is the one bit that would bother me. However, I claim it's thread safe as long as it's threadsafe to access `t.cancel` concurrently. It will never be `nil` for `Stop`.